addons/portal/controllers/mail.py

```diff
@@ -57,7 +57,7 @@ def _message_post_helper(res_model, res_id, message, token='', _hash=False, pid=
     # current user access rights (_mail_post_access use case).
     if token or (_hash and pid):
         pid = int(pid) if pid else False
-        if _check_special_access(res_model, res_id, token=token, _hash=_hash, pid=pid):
+        if _check_special_access(res_model, res_id, token=token if not pid else False, _hash=_hash, pid=pid):
```

Comment:
Shouldn't we instead change _check_special_access and do 1/ check on hash + pid (if given) 2/ check token (if given) 3/ raise ? I don't like messing with parameters depending on some context, it is better to give all parameters and let the function decide what to do with it.